src/client/editor/SceneLoader.js

```diff
@@ -64,41 +66,185 @@ function loadGLTF(url) {
+    const curShadowComponent = ShadowComponent.getComponent(object);
+    const curMaterialComponent = StandardMaterialComponent.getComponent(object);
+
+    if (object.userData._static && curShadowComponent && curMaterialComponent) {
+      let foundMaterial = false;
+
+      for (const { shadowComponent, materialComponent, meshes } of meshesToCombine) {
+        if (
+          shallowEquals(materialComponent.props, curMaterialComponent.props) &&
+          shallowEquals(shadowComponent.props, curShadowComponent.props)
+        ) {
+          meshes.push(object);
+          foundMaterial = true;
+          break;
         }
+      }
+
+      if (!foundMaterial) {
+        meshesToCombine.push({
+          shadowComponent: curShadowComponent,
+          materialComponent: curMaterialComponent,
+          meshes: [object]
+        });
+      }
+    }
+
+    // Remove objects marked as _dontExport
+    for (const child of object.children) {
+      if (child.userData._dontExport) {
+        object.remove(child);
+        return;
+      }
+    }
+  });
+
+  // Combine meshes and add to scene.
+  const invMatrix = new THREE.Matrix4();
+
+  for (const { meshes } of meshesToCombine) {
+    if (meshes.length > 1) {
+      const bufferGeometries = [];
+
+      for (const mesh of meshes) {
+        // Clone buffer geometry in case it is re-used across meshes with different materials.
+        const clonedBufferGeometry = mesh.geometry.clone();
+        invMatrix.getInverse(mesh.matrixWorld);
+        clonedBufferGeometry.applyMatrix(invMatrix);
+        bufferGeometries.push(clonedBufferGeometry);
+      }
+
+      const originalMesh = meshes[0];
+
+      const combinedGeometry = THREE.BufferGeometryUtils.mergeBufferGeometries(bufferGeometries);
+      delete combinedGeometry.userData.mergedUserData;
+      const combinedMesh = new THREE.Mesh(combinedGeometry, originalMesh.material);
+      combinedMesh.name = "CombinedMesh";
+      combinedMesh.receiveShadow = originalMesh.receiveShadow;
+      combinedMesh.castShadow = originalMesh.castShadow;
+
+      clonedScene.add(combinedMesh);
 
-        node.extras.components[component.name] = component.props;
+      for (const mesh of meshes) {
+        const meshIndex = mesh.parent.children.indexOf(mesh);
+        const parent = mesh.parent;
+        mesh.parent.remove(mesh);
+        const replacementObj = new THREE.Object3D();
+        replacementObj.copy(mesh, true);
+
+        addChildAtIndex(parent, replacementObj, meshIndex);
       }
     }
-    delete node.extras._components;
   }
-}
 
-function removeEditorData(node) {
-  if (node.extras) {
-    for (const key in node.extras) {
-      console.log(key);
+  const componentsToExport = Components.filter(c => !c.dontExportProps).map(component => component.componentName);
+
+  // Second pass at scene optimization.
+  clonedScene.traverse(object => {
+    const userData = object.userData;
+
+    // Move component data to userData.components
```

Comment:
Can we call it `spokeComponents` now?

Reply:
Probably not for the `userData` property. We want the components to line up with how Hubs expects them for now. We'll then probably want to move component data to individual extensions still or at minimum, namespaced extras fields.